### PR TITLE
PUBDEV-7020: extended demo for new target encoding API

### PR DESCRIPTION
**Pull request by deil87**
This PR is an extension of https://github.com/h2oai/h2o-3/pull/4474 and provides both docs with examples of how to use TE apis (Py/R) and equivalent tests that could be run and verified against.

Note: bug in new TE api was discovered but will be addressed in a dedicated PR

### CODE REVIEW

File: h2o-py/tests/testdir_algos/automl/target_encoding/pyunit_target_encoding_model.py [hunk 1]
```diff
@@ -62,6 +63,19 @@ def test_target_encoding_fit_method():
...
+                                             sample_rate=0.8,
+                                             col_sample_rate=0.8,
+                                             seed=1234,
+                                             stopping_rounds=5,
+                                             stopping_metric="AUC",
+                                             stopping_tolerance=0.001,
+                                             model_id="gbm_with_te")
+
+    myX = ["pclass", "sex", "age", "sibsp", "parch", "fare", "cabin_te", "embarked_te", "home.dest_te"]
+    gbm_with_te.train(x=myX, y=targetColumnName, training_frame=transformed)
```
Comment (deil87) [inline h2o-py/tests/testdir_algos/automl/target_encoding/pyunit_target_encoding_model.py:77]: @Pscheidl could you please take a look at this test? I added gbm estimator and trying to train it with new encodings. Training of the model fails when we are using frame which was encoded. Looks like frame is still being locked. May I be missing something?

Comment (Pscheidl) [reply h2o-py/tests/testdir_algos/automl/target_encoding/pyunit_target_encoding_model.py:77]: Yes, I'll try to help. Release duties today, so I'll have to postpone this to today evening CET or weekend.

Comment (deil87) [reply h2o-py/tests/testdir_algos/automl/target_encoding/pyunit_target_encoding_model.py:77]: created a jira https://0xdata.atlassian.net/browse/PUBDEV-7434

Comment (deil87) [reply h2o-py/tests/testdir_algos/automl/target_encoding/pyunit_target_encoding_model.py:77]: @Pscheidl did you have a chance to verify whether there is a bug?

Comment (Pscheidl) [reply h2o-py/tests/testdir_algos/automl/target_encoding/pyunit_target_encoding_model.py:77]: Not yet, but I can promise today.

Comment (Pscheidl) [reply h2o-py/tests/testdir_algos/automl/target_encoding/pyunit_target_encoding_model.py:77]: I pushed the fix. You can always revert if you don't like it, it modifies your code.

File: h2o-extensions/target-encoder/src/main/java/ai/h2o/targetencoding/TargetEncoder.java [hunk 2]
```diff
@@ -235,11 +235,15 @@ Frame imputeNAsForColumn(Frame data, String teColumnName, String strToImpute) {
...
-      DKV.put(updatedVec);
-      fr.update();
-      lock.unlock();
+      try {
+        Vec updatedVec = fr.vec(teColumnName);
+        updatedVec.setDomain(domain);
+        fr.update();
+      } finally {
+        lock.unlock();
+        fr.update();
```
Comment (michalkurka) [inline h2o-extensions/target-encoder/src/main/java/ai/h2o/targetencoding/TargetEncoder.java:244]: update after an `unlock`? seems suspicious

example of correct use: https://github.com/h2oai/h2o-3/blob/master/h2o-core/src/main/java/water/fvec/Frame.java#L1832

File: h2o-py/tests/testdir_algos/automl/target_encoding/pyunit_target_encoding_model.py (hunk 1 above)
Comment (deil87) [reply h2o-py/tests/testdir_algos/automl/target_encoding/pyunit_target_encoding_model.py:77]: @Pscheidl This PR is about demo and I think fix should be provided separately in the jira that I have created as it is a bug in a new version API. Same code is working with old API and there are tests that prove it (MichalK suggested me this piece of code and probably it is not incorrect, it just did not take into account your changes). Fix quite often could be placed into different places.

File: h2o-extensions/target-encoder/src/main/java/ai/h2o/targetencoding/TargetEncoder.java [hunk 3]
```diff
@@ -235,11 +235,15 @@ Frame imputeNAsForColumn(Frame data, String teColumnName, String strToImpute) {
...
     private void updateDomainGlobally(Frame fr, String teColumnName, String[] domain) {
       Lockable lock = fr.write_lock();
-      Vec updatedVec = fr.vec(teColumnName);
-      updatedVec.setDomain(domain);
-      DKV.put(updatedVec);
-      fr.update();
-      lock.unlock();
+      try {
+        Vec updatedVec = fr.vec(teColumnName);
+        updatedVec.setDomain(domain);
```
Comment (michalkurka) [inline h2o-extensions/target-encoder/src/main/java/ai/h2o/targetencoding/TargetEncoder.java:240]: don't drop the DKV.put - that is necessary for the changes to propagate to other nodes

File: h2o-py/tests/testdir_algos/automl/target_encoding/pyunit_target_encoding_model.py (hunk 1 above)
Comment (Pscheidl) [reply h2o-py/tests/testdir_algos/automl/target_encoding/pyunit_target_encoding_model.py:77]: I'll fix it in different PR then.

Comment (deil87) [reply h2o-py/tests/testdir_algos/automl/target_encoding/pyunit_target_encoding_model.py:77]: Thank you, @Pscheidl

File: h2o-extensions/target-encoder/src/main/java/ai/h2o/targetencoding/TargetEncoder.java (hunk 3 above)
Comment (Pscheidl) [reply h2o-extensions/target-encoder/src/main/java/ai/h2o/targetencoding/TargetEncoder.java]: It's just the update propagation problem to me.

Comment (deil87) [reply h2o-extensions/target-encoder/src/main/java/ai/h2o/targetencoding/TargetEncoder.java]: @Pscheidl yeah, maybe. Did you have a chance to try this fix in a separate PR, to see results from mr Jenkins?

Comment (Pscheidl) [reply h2o-extensions/target-encoder/src/main/java/ai/h2o/targetencoding/TargetEncoder.java]: Creating it right now, I'm investigating the lock mechanism for obvious issues. I need approx. 30 minutes.

Comment (Pscheidl) [reply h2o-extensions/target-encoder/src/main/java/ai/h2o/targetencoding/TargetEncoder.java]: Here is the continuation: https://github.com/h2oai/h2o-3/pull/4501

File: h2o-extensions/target-encoder/src/main/java/ai/h2o/targetencoding/TargetEncoder.java (hunk 2 above)
Comment (Pscheidl) [reply h2o-extensions/target-encoder/src/main/java/ai/h2o/targetencoding/TargetEncoder.java]: https://github.com/h2oai/h2o-3/pull/4501

**Comment by deil87** [comment]
@angela0xdata tests are fine and probably we are ready to merge this into your PR. You can try demo code for yourself as well. Let me know if I can merge it into your PR.

**Comment by angela0xdata** [comment]
Thanks @deil87. I think this can be merged now.

**Pull request merged**